rtCommon/bidsArchive.py

```diff
@@ -0,0 +1,825 @@
+        """
+        Updates the layout of the dataset so that any new metadata or image
+        files are added to the index.
+        """
+        self.data = BIDSLayout(self.rootPath)
+
+    def _addImage(self, img: nib.Nifti1Image, path: str) -> None:
+        """
+        Add an image to the dataset at the provided path, creating the path if
+        it does not yet exist.
+        """
+        logger.debug("Writing new image to %s", self.absPathFromRelPath(path))
+        layoutUpdateRequired = self._ensurePathExists(os.path.dirname(path))
+
+        absPath = self.absPathFromRelPath(path)
+        nib.save(img, absPath)
+
+        if layoutUpdateRequired:
+            self._updateLayout()
+
+    def _addMetadata(self, metadata: dict, path: str) -> None:
+        absPath = self.absPathFromRelPath(path)
+
+        with open(absPath, 'w', encoding='utf-8') as metadataFile:
+            json.dump(metadata, metadataFile, ensure_ascii=False, indent=4)
+
+        logger.debug("Wrote new metadata to %s", absPath)
+
+        self._updateLayout()
+
+    # Used to update the archive if any on-disk changes have happened
+    def _update(self):
+        if self.data:
+            self._updateLayout()
+
+    def isEmpty(self) -> bool:
+        return (self.data is None)
+
+    def tryGetFile(self, path: str) -> BIDSFile:
+        """
+        Tries to get a file from the archive using different interpretations of
+        the target path. Interpretations considered are:
+        1) Path with leading slash, relative to filesystem root
+        2) Path with leading slash, relative to archive root
+        3) Path with no leading slash
+
+        Args:
+            path: Path to the file to attempt to get.
+
+        Returns:
+            BIDSFile (or subclass) if a matching file was found, None otherwise.
+
+        Examples:
+            >>> archive = BidsArchive('/path/to/archive')
+            >>> filename = 'sub-01_task-test_bold.nii.gz'
+            >>> archive.tryGetFile('/tmp/archive/sub-01/func/' + filename)
+            <BIDSImageFile filename=/tmp/archive/sub-01/func/sub-01_task-test\
+            _bold.nii.gz
+            >>> archive.tryGetFile('/' + filename)
+            <BIDSImageFile filename=/tmp/archive/sub-01/func/sub-01_task-test\
+            _bold.nii.gz
+            >>> archive.tryGetFile(filename)
+            <BIDSImageFile filename=/tmp/archive/sub-01/func/sub-01_task-test\
+            _bold.nii.gz
+        """
+        # Tries absolute to disk root and relative to archive root
+        archiveFile = self.get_file(path)
+        if archiveFile is not None:
+            return archiveFile
+
+        # Try absolute to archive root
+        strippedRootPath = self._stripLeadingSlash(path)
+        archiveFile = self.get_file(strippedRootPath)
+        if archiveFile is not None:
+            return archiveFile
+
+        return None
+
+    @failIfEmpty
+    def getMetadata(self, path: str, includeEntities: bool = True) -> dict:
```

Comment:
GW comment: I wonder if this should be called getSidecarMetadata()? so as not to confuse it with getting the image header metadata.